### PR TITLE
ENH: io.wavfile: read unseekable files

### LINKED ISSUE #11328: Scipy unable to read piped wav file
Scipy version 1.4.1

Repro:
```
import subprocess
import scipy
sp = subprocess.Popen(
    ['ffmpeg', '-i', 'test.mp3', '-f', 'wav', '-ac', '1', '-ar', '22050', '-'],
    stdout=subprocess.PIPE)
scipy.io.wavfile.read(sp.stdout)
```

Gives:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/scipy/io/wavfile.py", line 273, in read
    while fid.tell() < file_size:
OSError: [Errno 29] Illegal seek
```

### PR DESCRIPTION
**Pull request by smammy**
#### Reference issue
Closes gh-11328

#### What does this implement/fix?
Allow scipy.io.wavfile.read() to read non-seekable files.

#### Additional information
Allow scipy.io.wavfile.read() to read non-seekable files.

  * If passed stream is not seekable, wrap it in a SeekEmulatingReader, which tracks current position and implements tell() to report it.

Most seek()s in wavfile.read() simply seek forward from the current position by a fixed amount. To address these:

  * Support limited SeekEmulatingReader.seek() that emulates seek()s that can be supported by read()ing and discarding the result.

Befor returning, read() attempts to seek back to the start of the file. This is nice when possible, but if it isn't we shouldn't blow up. So:

  * Avoid final fid.seek(0) if underlying stream isn't seekable.

The remaining seek()s are in RF64 support, where chunk size is global and stored in the RIFF header rather than in the chunk header. _read_data_chunk() currently seeks back to a fixed offset from the file origin to grab the chunk size. To address this:

  * Modify RF64 support to parse and store the chunk size in _read_riff_chunk() so that _read_data_chunk() doesn't have to seek() back for it. This applies whether or not reading from a seekable stream.

To test this, I added test_streams(), which ensures that wavfile reads the same data no matter whether a path, seekable stream, or unseekable stream is passed to read().

### CODE REVIEW

**Comment by j-bowhay** [comment]
@hpx7 would you be able to have a look, does this fix your issue?

**Comment by j-bowhay** [comment]
Thanks again @nickodell for reviewing #22215, would you have time to have a look at this too?

File: scipy/io/tests/test_wavfile.py
Comment (nickodell) [inline scipy/io/tests/test_wavfile.py:1]: GitHub says the mode on these files has changed from 644 to 755. Since these files are not intended to be executed, can you change them back?

File: scipy/io/wavfile.py [hunk 1]
```diff
@@ -27,6 +28,39 @@ class WavFileWarning(UserWarning):
...
+    def seek(self, offset, whence=os.SEEK_SET, /):
+        match whence:
+            case os.SEEK_SET if offset == self.pos:
+                pass # advance to current position (no-op)
+            case os.SEEK_CUR if offset >= 0:
+                self.read(offset) # advance by offset
+            case os.SEEK_END if offset == 0:
+                self.read() # advance to end of stream
+            case _:
+                raise io.UnsupportedOperation()
```
Comment (nickodell) [inline scipy/io/wavfile.py:50]: This should have an exception message to clarify why it's being thrown.

Here's some example language:

>"SeekEmulatingReader was asked to emulate a seek operation it does not support."

Since users aren't expected to trigger it, it's okay if it's fairly technical.

File: scipy/io/wavfile.py [hunk 2]
```diff
@@ -27,6 +28,39 @@ class WavFileWarning(UserWarning):
...
+    
+    def tell(self):
+        return self.pos
+    
+    def close(self):
+        self.reader.close()
+    
+    def flush(self):
+        # make np.fromfile fail intelligibly
+        raise io.UnsupportedOperation()
```
Comment (nickodell) [inline scipy/io/wavfile.py:61]:
```
# make np.fromfile fail intelligibly
```

Can you elaborate? What is this protecting against?

File: scipy/io/wavfile.py [hunk 3]
```diff
@@ -728,7 +765,10 @@ def read(filename, mmap=False):
         if not hasattr(filename, 'read'):
             fid.close()
         else:
-            fid.seek(0)
```
Comment (nickodell) [inline scipy/io/wavfile.py:731]: I am frankly mystified as to why this final seek operation was originally here. Most other libraries will not seek to the beginning of a file if you hand them an open file to read - e.g. if you use `pd.read_csv()` to read a file, the file position after the function call will be where it stopped reading.

I would suggest removing `fid.seek(0)`, and not seeking even if the file supports it.

Thoughts, @j-bowhay ?

File: scipy/io/wavfile.py (hunk 2 above)
Comment (nickodell) [reply scipy/io/wavfile.py:61]: Also, this should have an exception message to clarify why it's being thrown.

**Review by nickodell — CHANGES_REQUESTED** [review]
I like the general concept of this change, but there are a few details that need to be addressed.

**Comment by nickodell** [comment]
I wrote a test script for this PR to check that it works when reading wav data from a pipe, and it does.

Example:

```python
import scipy
#import gzip
#import io
import subprocess

def checksum_first_channel(res):
    sr, a = res
    if a.ndim == 2:
        chk = a[:, 0].astype('float').sum()
    elif a.ndim == 1:
        chk = a.astype('float').sum()
    else:
        raise Exception()
    print("chk", chk)
    return chk

def read_from_subproc(filename):
    sp = subprocess.Popen(
        ['cat', filename],
        stdout=subprocess.PIPE
    )
    return sp.stdout

a = scipy.io.wavfile.read('../wavtest1-rf64.wav')
assert checksum_first_channel(a) == -122419968.0
a = scipy.io.wavfile.read('../wavtest2.wav')
assert checksum_first_channel(a) == -82594741.0

a = scipy.io.wavfile.read(read_from_subproc('../wavtest1-rf64.wav'))
assert checksum_first_channel(a) == -122419968.0

a = scipy.io.wavfile.read(read_from_subproc('../wavtest2.wav'))
assert checksum_first_channel(a) == -82594741.0
```

I checked this with PCM and RF64 wav files.

File: scipy/io/tests/test_wavfile.py
Comment (nickodell) [reply scipy/io/tests/test_wavfile.py:1]: (This applies to both files.)

File: scipy/io/wavfile.py (hunk 3 above)
Comment (j-bowhay) [reply scipy/io/wavfile.py:731]: It was added in 47b9012, @nils-werner do you happen to remember?

Comment (nils-werner) [reply scipy/io/wavfile.py:731]: If I remember correctly it was there before, I just proposed the `if` clause. Maybe follow the `git blame` to the changes before?

I don't think it needs to be the at all tbh.

Comment (smammy) [reply scipy/io/wavfile.py:731]: `seek(0)` doesn't seem to appear prior to 47b9012, maybe it had something to do with StringIO compatibility?

Removing it makes sense to me, but I worry it could break user code that depends on that behavior.

Comment (nils-werner) [reply scipy/io/wavfile.py:731]: Ah right, IIRC that was to fill the buffer with RIFF data, and then rewind it so that the buffer can immediately be used by something that wants to read a WAV file (which contains RIFF data).

The use case was the `Audio` tag in Jupiter, where you'd want to export an array to a WAV file-like buffer, and then send it over to the frontend.

Comment (smammy) [reply scipy/io/wavfile.py:731]: Ah ok, that makes sense. I've added a comment.

**Comment by smammy** [comment]
Thanks for the review! I've pushed changes, and can squash them if you like.

File: scipy/io/wavfile.py (hunk 3 above)
Comment (nickodell) [reply scipy/io/wavfile.py:731]: Commit dd3568a34f6a19b67755b93934e910c55f9dbde2 looks like a nice compromise: it does not swallow exceptions if the underlying stream claims to support seeking, but actually can't. This was my primary concern. It also doesn't change functionality in case applications are relying on the `seek(0)` behavior.

File: scipy/io/wavfile.py (hunk 2 above)
Comment (nickodell) [reply scipy/io/wavfile.py]: I've investigated this in a little more detail.

The reason why `np.fromfile()` calls `flush()` is that it could be reading from a read/write file handle, where changes were just made. While reading, it flushes changes to the file, duplicates the file handle, seeks that handle to the offset, then starts reading. It closes the duplicated file handle. See https://github.com/numpy/numpy/issues/7831 for detail.

This has two implications.

1. First, reading from unseekable streams will never use the `np.fromfile()` fast path, because it calls `flush()`.
2. Second, even if we added flush support, e.g.

   ```python
   def flush(self):
       return self.reader.flush()
   ```

   this would still not allow use of the fast path, because NumPy wants to seek the stream, which this stream does not support. We cannot intercept this seek, because it is done in C. NumPy does this seek even if offset=0, so we cannot avoid it. [Link to code](https://github.com/numpy/numpy/blob/dfb00057a5bba94284dfbebffe37c37b08a16d42/numpy/_core/src/multiarray/multiarraymodule.c#L2398)

In summary, throwing an exception is the best we can do here.

File: scipy/io/tests/test_wavfile.py
Comment (nickodell) [reply scipy/io/tests/test_wavfile.py:1]: This concern was addressed.

File: scipy/io/wavfile.py (hunk 1 above)
Comment (nickodell) [reply scipy/io/wavfile.py]: This concern was addressed.

**Comment by nickodell** [comment]
> Thanks for the review! I've pushed changes, and can squash them if you like.

Yes, please do.

**Review by nickodell — APPROVED** [review]

File: scipy/io/wavfile.py (hunk 2 above)
Comment (smammy) [reply scipy/io/wavfile.py]: Thanks for chasing that down. And there are additional seeks that can't be avoided without changing numpy's API, e.g. [`array_fromfile_binary`](https://github.com/numpy/numpy/blob/965fdf523fa66cdaf37943f28a6a14f7ae91198e/numpy/_core/src/multiarray/ctors.c#L3480)` in `ctors.c`.

**Comment by smammy** [comment]
Squashed, and split a long string to make the linter happy.

**Comment by smammy** [comment]
(Ugh, sorry for the multiple force-pushes, there were stray whitespace changes that I fumbled. The change for the linter [looks like this](https://github.com/scipy/scipy/compare/dd3568a34f6a19b67755b93934e910c55f9dbde2..96d162db5f38604096ea5eaa966b8320933c56d9).)

File: scipy/io/tests/test_wavfile.py [hunk 4]
```diff
@@ -334,6 +334,47 @@ def test_write_roundtrip_rf64(tmpdir):
...
+                     'test-8000Hz-le-3ch-5S-53bit.wav',
+                     'test-8000Hz-le-3ch-5S-64bit.wav',
+                     'test-44100Hz-be-1ch-4bytes.wav', # RIFX
+                     'test-44100Hz-le-1ch-4bytes-rf64.wav']:
+        dfname = datafile(filename)
+        with open(dfname, 'rb') as fp1, open(dfname, 'rb') as fp2:
+            rate1, data1 = wavfile.read(fp1)
+            rate2, data2 = wavfile.read(Nonseekable(fp2))
+            rate3, data3 = wavfile.read(dfname, mmap=False)
+            assert_array_equal(rate1, rate3)
```
Comment (j-bowhay) [inline scipy/io/tests/test_wavfile.py:372]: Nit, could you change the new tests to use `numpy.testing.assert_equal` as per the recommendations in the numpy docs

File: scipy/io/tests/test_wavfile.py [hunk 5]
```diff
@@ -334,6 +334,47 @@ def test_write_roundtrip_rf64(tmpdir):
...
+            assert_array_equal(rate2, rate3)
+            assert_array_equal(data1, data3)
+            assert_array_equal(data2, data3)
```
Comment (j-bowhay) [inline scipy/io/tests/test_wavfile.py:375]:
```suggestion
            assert_equal(rate1, rate3)
            assert_equal(rate2, rate3)
            assert_equal(data1, data3)
            assert_equal(data2, data3)
```

**Comment by j-bowhay** [comment]
Thanks for a great first contribution to SciPy @smammy, please feel free to open prs for anything else you might be interested in working on:)

Thanks as ever for the thorough review @nickodell

**Comment by smammy** [comment]
Thanks everyone for your help and review! Glad to learn something and get this working.
